### ISSUE TITLE
twisted.web.html is old, crufty, and XSS-vulnerable

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4948]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)| @habnabit reported|
|-|-|
|Trac ID|trac#4948|
|Type|defect|
|Created|2011-03-14 19:32:09Z|
|Branch|https://github.com/twisted/twisted/tree/webhtml-4948-4|

The current contents of twisted.web.html needs to be deprecated. Places that use it need to be changed to instead use twisted.web.template (see [#4939](https://github.com/twisted/twisted/issues/4939)).

Most of this code doesn't bother with escaping, and therefore introduces potential XSS vulnerabilities. This is a Very Bad Thing, of course. twisted.web.template will fix this by virtue of being used at all.

<details><summary>Searchable metadata</summary>

```
trac-id__4948 4948
type__defect defect
reporter__habnabit habnabit
priority__normal normal
milestone__ 
branch__branches_webhtml_4948_4 branches/webhtml-4948-4
branch_author__adiroiban__habnabit adiroiban, habnabit
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1300131129000000 1300131129000000
changetime__1436855296392191 1436855296392191
version__None None
owner__adiroiban adiroiban
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_2"></a>|@habnabit commented|
|-|-|

(In [31043]) Branching to 'webhtml-4948'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit removed owner<br>@habnabit set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @habnabit<br>@exarkun set status to `new`|
|-|-|

Thanks!

  1. Needs unit tests!  Take a look at, eg, `test_moduleMovedForSplitDeprecation` in `twisted/python/test/test_util.py`
  1. There are some failing unit tests in `twisted.web.test` caused by this change.
  1. There are some uses in Twisted itself of these deprecated APIs.  Twisted shouldn't trip over its own deprecations.
    1. `twisted.web.util.formatFailure` is the main one, but there might be others (once the failing web tests run it should be easier to find them)
  1. Add a `.removal` [ReviewProcess#Newsfiles news fragment]

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Oops.  There appears to be some overlap between this ticket and [#4896](https://github.com/twisted/twisted/issues/4896).

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban commented|
|-|-|

(In [45080]) Branching to webhtml-4948-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban set owner to @adiroiban|
|-|-|

let me try to do this

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban commented|
|-|-|

First we need to remove usage to twisted.web.html . I have created a separate ticket [#7942](https://github.com/twisted/twisted/issues/7942)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban removed owner|
|-|-|

Current branch should not be merged before [[#7942](https://github.com/twisted/twisted/issues/7942)](https://github.com/twisted/twisted/issues/7942) but I and sending it for review as the twisted.web.html part is done...

once #7942 is merged , test should pass on this branch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban commented|
|-|-|

(In [45134]) Branching to webhtml-4948-3.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_14"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this.

The deprecation warnings don't say to use twisted.web.template instead. If you do  ```@deprecated(Version('Twisted', 15, 3, 0), replacement="twisted.web.template")```, then the deprecation warning should be correct.

Please fix that issue, rerun the builders, and merge when they're happy.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_15"></a>|@adiroiban commented|
|-|-|

(In [45253]) Branching to webhtml-4948-4.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_16"></a>|@adiroiban commented|
|-|-|

Thanks for the review. I have updated the branch with hint about the replacement.

some builders are failing as they are now on lxc :(

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_17"></a>|@adiroiban set status to `closed`|
|-|-|

(In [45255]) Merge webhtml-4948-4: Deprecate twisted.web.html.

Author: adiroiban
Reviewers: hawkowl
Fixes: [#4948](https://github.com/twisted/twisted/issues/4948)
